### PR TITLE
WAGE: MacText integration

### PR DESCRIPTION
**Pull request by VelocityRa**
**Edit:** I have fixed the bad commits, but this should still not be merged yet, it's not complete.

I'm submitting it now because it's my GSoC qualification PR.

### CODE REVIEW

File: graphics/macgui/mactext.cpp [hunk 1]
```diff
@@ -30,7 +30,7 @@ const Font *MacFontRun::getFont() {
 	if (font)
 		return font;
 
-	MacFont macFont = MacFont(fontId, fontSize, textSlant);
+	MacFont macFont = MacFont();
```
Comment (sev-) [inline graphics/macgui/mactext.cpp]: Why this was removed?

File: engines/wage/gui.cpp [hunk 2]
```diff
@@ -172,11 +179,37 @@ Gui::Gui(WageEngine *engine) {
...
+	//TODO: Make the font we use here work
+	// (currently MacFontRun::getFont gets called with the fonts being uninitialized,
+	// so it initializes them by itself with default params, and not those here)
+#ifdef USE_NEW_TEXT_RENDERER
+#ifdef USE_MACTEXTWINDOW
+	const Graphics::Font *font = _wm._fontMan->getFont(Graphics::MacFont(Graphics::kMacFontChicago, 8));
+
+	unsigned maxWidth = _screen.w;
+
+	 _consoleWindow = new Graphics::MacTextWindow(&_wm, font, kColorBlack, kColorWhite,
```
Comment (sev-) [inline engines/wage/gui.cpp]: Extra indentation space.

File: engines/wage/gui.cpp [hunk 3]
```diff
@@ -172,11 +179,37 @@ Gui::Gui(WageEngine *engine) {
...
 	_consoleWindow->setCallback(consoleWindowCallback, this);
 
 	loadBorders();
 
+#ifdef USE_NEW_TEXT_RENDERER
+#ifndef USE_MACTEXTWINDOW
+	unsigned maxWidth = _screen.w;
+
+	_mactext = new Graphics::MacText("", &_wm, font,
+		+kColorBlack, kColorWhite, maxWidth, Graphics::kTextAlignCenter);
```
Comment (sev-) [inline engines/wage/gui.cpp]: What is "+kColorBlack"?

File: graphics/macgui/mactextwindow.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "macwindowmanager.h"
```
Comment (sev-) [inline graphics/macgui/mactextwindow.cpp]: This is wrong. Full paths must be used.

File: graphics/macgui/mactextwindow.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef GRAPHICS_MACGUI_MACTEXTWINDOW_H
+#define GRAPHICS_MACGUI_MACTEXTWINDOW_H
+
+#include "mactext.h"
```
Comment (sev-) [inline graphics/macgui/mactextwindow.h]: Full paths only, please.

File: graphics/macgui/macwindowmanager.h [hunk 6]
```diff
@@ -104,6 +104,17 @@ class MacWindowManager {
 	 */
 	MacWindow *addWindow(bool scrollable, bool resizable, bool editable);
 	/**
+	 * Adds a window that has already been initialized to the registry.
+	 * Like addWindow, but this doesn't create/allocate the Window.
+	 * @param macWindow the window to be added to the registry
+	 */
+	void MacWindowManager::addWindowInitialized(MacWindow *macwindow);
```
Comment (sev-) [inline graphics/macgui/macwindowmanager.h]: This doesn't even compile on non-MSVC. See Travis CI logs.

File: engines/wage/gui.cpp [hunk 7]
```diff
@@ -172,11 +179,37 @@ Gui::Gui(WageEngine *engine) {
...
 	_sceneWindow->setCallback(sceneWindowCallback, this);
 
+	//TODO: Make the font we use here work
+	// (currently MacFontRun::getFont gets called with the fonts being uninitialized,
+	// so it initializes them by itself with default params, and not those here)
+#ifdef USE_NEW_TEXT_RENDERER
+#ifdef USE_MACTEXTWINDOW
+	const Graphics::Font *font = _wm._fontMan->getFont(Graphics::MacFont(Graphics::kMacFontChicago, 8));
+
+	unsigned maxWidth = _screen.w;
```
Comment (sev-) [inline engines/wage/gui.cpp]: unsigned what? Please use full type identifiers.

File: graphics/macgui/mactext.cpp [hunk 8]
```diff
@@ -324,20 +324,33 @@ void MacText::draw(ManagedSurface *g, int x, int y, int w, int h, int xoff, int
...
-	g->blitFrom(*_surface, Common::Rect(MIN<int>(_surface->w, x),     MIN<int>(_surface->h, y),
-									    MIN<int>(_surface->w, x + w), MIN<int>(_surface->w, y + w)),
+	g->blitFrom(*_surface, Common::Rect(MIN<int>(_surface->w, x), MIN<int>(_surface->h, y),
+									    MIN<int>(_surface->w, x + w), MIN<int>(_surface->h, y + w)),
 										Common::Point(xoff, yoff));
 }
 
 void MacText::appendText(Common::String str) {
-	int oldLen = _textLines.size();
+	unsigned oldLen = _textLines.size();
```
Comment (sev-) [inline graphics/macgui/mactext.cpp]: please use full type identifiers. Like 'uint'

File: graphics/macgui/macwindowmanager.cpp [hunk 9]
```diff
@@ -178,20 +178,27 @@ MacWindow *MacWindowManager::addWindow(bool scrollable, bool resizable, bool edi
...
 	return w;
 }
 
+void MacWindowManager::addWindowInitialized(MacWindow *macwindow) {
+	_windows.push_back(macwindow);
+	_windowStack.push_back(macwindow);
+}
+
+int MacWindowManager::getNextId() {
+	auto lastId = _lastId;
```
Comment (sev-) [inline graphics/macgui/macwindowmanager.cpp]: Please, do not use auto variables.

**Comment by VelocityRa** [comment]
Thank you, I will handle all of those issues later when I get home.

**Comment by VelocityRa** [comment]
I cleaned things up a bit for someone else to continue the integration.
I can answer any questions here or in IRC.

To use the new text rendering uncomment the `USE_MACTEXTWINDOW` define in `gui.h`.

Main things to do left are text selection/copy/paste and input rendering. These should be ported over from `gui-console.cpp`.

**Pull request merged**